firedrake/preconditioners/fdm.py

```diff
@@ -70,197 +44,267 @@ class FDMPC(PCBase):
         # however from the PETSc command line.  Since PC allows the user to specify
         # a KSP, we can do iterative by -fdm_pc_type ksp.
-        pc = PETSc.PC().create(comm=opc.comm)
-        pc.incrementTabLevel(1, parent=opc)
-
-        # We set a DM on the constructed PC so one
-        # can do patch solves with ASMPC.
-        dm = opc.getDM()
-        pc.setDM(dm)
-        pc.setOptionsPrefix(options_prefix)
-        pc.setOperators(self.Pmat, self.Pmat)
-        self.pc = pc
-        pc.setFromOptions()
-        self.update(pc)
+        fdmpc = PETSc.PC().create(comm=pc.comm)
+        fdmpc.incrementTabLevel(1, parent=pc)
 
-    def update(self, pc):
-        for assemble_coef in self.assembly_callables:
-            assemble_coef()
-        self.Pmat.zeroEntries()
-        self._assemble_Pmat()
-        self.Pmat.zeroRowsColumnsLocal(self.bc_nodes)
-
-    def apply(self, pc, x, y, transpose=False):
-        with self.uc.dat.vec_wo as xc:
-            xc.set(0.0E0)
-
-        with self.uf.dat.vec_wo as xf:
-            x.copy(xf)
-
-        op2.par_loop(self.restrict_kernel, self.mesh.cell_set,
-                     self.uc.dat(op2.INC, self.cell_node_map),
-                     self.uf.dat(op2.READ, self.cell_node_map),
-                     self.weight.dat(op2.READ, self.cell_node_map))
-
-        for bc in self.bcs:
-            bc.zero(self.uc)
-
-        with self.uc.dat.vec_ro as x_, self.uf.dat.vec_wo as y_:
-            if transpose:
-                self.pc.applyTranspose(x_, y_)
-            else:
-                self.pc.apply(x_, y_)
+        # We set a DM and an appropriate SNESContext on the constructed PC so one
+        # can do e.g. multigrid or patch solves.
+        fdm_dm = V_fdm.dm
+        self._dm = fdm_dm
 
-        for bc in self.bcs:
-            bc.zero(self.uf)
+        fdmpc.setDM(fdm_dm)
+        fdmpc.setOptionsPrefix(options_prefix)
+        fdmpc.setOperators(A=Amat, P=Pmat)
+        fdmpc.setUseAmat(True)
+        self.pc = fdmpc
 
-        op2.par_loop(self.prolong_kernel, self.mesh.cell_set,
-                     self.uc.dat(op2.WRITE, self.cell_node_map),
-                     self.uf.dat(op2.READ, self.cell_node_map))
+        with dmhooks.add_hooks(fdm_dm, self, appctx=self._ctx_ref, save=False):
+            fdmpc.setFromOptions()
+        # self.diagonal_scaling()
 
-        with self.uc.dat.vec_ro as xc:
-            xc.copy(y)
-
-        y.array_w[self.bc_nodes] = x.array_r[self.bc_nodes]
+    @PETSc.Log.EventDecorator("FDMUpdate")
+    def update(self, pc):
+        if hasattr(self, "A"):
+            self._assemble_A()
+        self._assemble_P()
+        # self.diagonal_scaling()
+
+    def apply(self, pc, x, y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            if hasattr(self, "fdm_interp"):
+                with self.work.dat.vec as x_fdm, self.diag.dat.vec as y_fdm:
```

Comment:
both of these accesses can be `vec_wo` I think. Or better, just keep `Vec`s hanging around, rather than `Function`s?

Reply:
Good suggestion. I'll go for the `Vec` approach.